### PR TITLE
WebServer+LibHTTP: Make WebServer more robust and add HTTP basic authentication support

### PR DESCRIPTION
**Pull request by MaxWipfli**
This fixes various bugs in WebServer. It also adds support for HTTP basic authentication to it, which can be enabled with the `--username` and `--password` command line flags.

WebServer now works much better than before, but the most severe issue remains: If the client is too slow in sending the request, everything blows up. See #7880.

### CODE REVIEW

File: Userland/Libraries/LibHTTP/HttpResponse.cpp [hunk 1]
```diff
@@ -18,4 +18,63 @@ HttpResponse::~HttpResponse()
...
+        { 408, "Request Timeout"sv },
+        { 409, "Conflict"sv },
+        { 410, "Gone"sv },
+        { 411, "Length Required"sv },
+        { 412, "Precondition Failed"sv },
+        { 413, "Payload Too Large"sv },
+        { 414, "URI Too Long"sv },
+        { 415, "Unsupported Media Type"sv },
+        { 416, "Range Not Satisfiable"sv },
+        { 417, "Expectation Failed"sv },
```
Comment (stelar7) [inline Userland/Libraries/LibHTTP/HttpResponse.cpp:59]: You cant add status codes, and not include `418 I'm a teapot` :^)

Comment (MaxWipfli) [reply Userland/Libraries/LibHTTP/HttpResponse.cpp:59]: It's sadly not part of the RFC, and I decided to stay with that list for the moment. :^)

Comment (awesomekling) [reply Userland/Libraries/LibHTTP/HttpResponse.cpp:59]: Thank you for not memeing in the code <3

Comment (MaxWipfli) [reply Userland/Libraries/LibHTTP/HttpResponse.cpp:59]: > Thank you for not memeing in the code <3

Although it is actually listed by [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses).

File: Userland/Services/WebServer/Client.cpp [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include <LibCore/FileStream.h>
```
Comment (awesomekling) [inline Userland/Services/WebServer/Client.cpp:19]: Typo in commit message, canoical -> canonical

File: Userland/Services/WebServer/Client.cpp [hunk 3]
```diff
@@ -217,15 +217,21 @@ void Client::handle_directory_listing(String const& requested_path, String const
         StringBuilder path_builder;
         path_builder.append(real_path);
         path_builder.append('/');
-        path_builder.append(name);
```
Comment (gunnarbeutner) [inline Userland/Services/WebServer/Client.cpp:213]: Commit message: `evaulautes` -> `evaluates`. :)

File: Userland/Libraries/LibHTTP/HttpResponse.cpp [hunk 4]
```diff
@@ -18,4 +18,63 @@ HttpResponse::~HttpResponse()
 {
 }
 
+StringView HttpResponse::reason_phrase_for_code(int code)
+{
+    VERIFY(code > 0);
+
+    static HashMap<int, String> s_reason_phrases = {
```
Comment (linusg) [inline Userland/Libraries/LibHTTP/HttpResponse.cpp:25]: We're constructing a bunch of `StringView`s, putting them in a `HashMap` of `String`s and convert them back to `StringView`s when returning - Is that actually needed?
Also, can this and the function itself be `constexpr`?

File: Userland/Libraries/LibHTTP/HttpResponse.cpp [hunk 5]
```diff
@@ -18,4 +18,63 @@ HttpResponse::~HttpResponse()
 {
 }
 
+StringView HttpResponse::reason_phrase_for_code(int code)
+{
+    VERIFY(code > 0);
```
Comment (linusg) [inline Userland/Libraries/LibHTTP/HttpResponse.cpp:23]:
```suggestion
    VERIFY(code >= 100);
    VERIFY(code <= 599);
```

File: Userland/Libraries/LibHTTP/HttpResponse.cpp (hunk 4 above)
Comment (MaxWipfli) [reply Userland/Libraries/LibHTTP/HttpResponse.cpp:25]: It cannot be `constexpr` because `HashMap` does not support it.

**Pull request merged**